pkg/util/diff/diff.go

```diff
@@ -74,6 +77,130 @@ func ObjectGoPrintDiff(a, b interface{}) string {
 	)
 }
 
+func ObjectReflectDiff(a, b interface{}) string {
+	vA, vB := reflect.ValueOf(a), reflect.ValueOf(b)
+	if vA.Type() != vB.Type() {
+		return fmt.Sprintf("type A %T and type B %T do not match", a, b)
+	}
+	diffs := objectReflectDiff(field.NewPath("object"), vA, vB)
+	if len(diffs) == 0 {
+		return ""
+	}
+	out := []string{""}
+	for _, d := range diffs {
+		out = append(out,
+			fmt.Sprintf("%s:", d.path),
+			limit(fmt.Sprintf("  a: %#v", d.a), 80),
+			limit(fmt.Sprintf("  b: %#v", d.b), 80),
+		)
+	}
+	return strings.Join(out, "\n")
+}
+
+func limit(s string, max int) string {
+	if len(s) > max {
+		return s[:max]
+	}
+	return s
+}
+
+func public(s string) bool {
+	if len(s) == 0 {
+		return false
+	}
+	return s[:1] == strings.ToUpper(s[:1])
+}
+
+type diff struct {
+	path *field.Path
+	a, b interface{}
+}
+
+func objectReflectDiff(path *field.Path, a, b reflect.Value) []diff {
+	switch a.Type().Kind() {
+	case reflect.Struct:
+		var changes []diff
+		for i := 0; i < a.Type().NumField(); i++ {
+			if !public(a.Type().Field(i).Name) {
+				if reflect.DeepEqual(a.Interface(), b.Interface()) {
+					return nil
+				}
+				return []diff{{path: path, a: fmt.Sprintf("%#v", a), b: fmt.Sprintf("%#v", b)}}
+			}
+			if sub := objectReflectDiff(path.Child(a.Type().Field(i).Name), a.Field(i), b.Field(i)); len(sub) > 0 {
+				changes = append(changes, sub...)
+			}
+		}
+		return changes
+	case reflect.Ptr:
+		if a.IsNil() || b.IsNil() {
+			switch {
+			case a.IsNil() && b.IsNil():
+				return nil
+			case a.IsNil():
+				return []diff{{path: path, a: nil, b: b.Interface()}}
+			default:
+				return []diff{{path: path, a: a.Interface(), b: nil}}
+			}
+		}
+		return objectReflectDiff(path, a.Elem(), b.Elem())
+	case reflect.Chan:
+		if !reflect.DeepEqual(a.Interface(), b.Interface()) {
+			return []diff{{path: path, a: a.Interface(), b: b.Interface()}}
+		}
+		return nil
+	case reflect.Slice:
+		if reflect.DeepEqual(a, b) {
+			return nil
+		}
+		lA, lB := a.Len(), b.Len()
+		l := lA
+		if lB < lA {
+			l = lB
+		}
+		for i := 0; i < l; i++ {
+			if !reflect.DeepEqual(a.Index(i), b.Index(i)) {
+				return objectReflectDiff(path.Index(i), a.Index(i), b.Index(i))
+			}
+		}
+		if lA > l {
```

Comment:
Going to fix it real quick (the reflect part).

On Mon, May 23, 2016 at 10:10 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/util/diff/diff.go
> <https://github.com/kubernetes/kubernetes/pull/25983#discussion_r64228825>
> :
>
> > +		return nil
> > +	case reflect.Slice:
> > +		if reflect.DeepEqual(a, b) {
> > +			return nil
> > +		}
> > +		lA, lB := a.Len(), b.Len()
> > +		l := lA
> > +		if lB < lA {
> > +			l = lB
> > +		}
> > +		for i := 0; i < l; i++ {
> > +			if !reflect.DeepEqual(a.Index(i), b.Index(i)) {
> > +				return objectReflectDiff(path.Index(i), a.Index(i), b.Index(i))
> > +			}
> > +		}
> > +		if lA > l {
>
> Sure. Maybe a TODO?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25983/files/7b51bc6a70c1d5d0b2e843e85334040ea8347c06..3d793f470ccf0f24782fd2fa2671c6da43e307e5#r64228825>
>